File: python/day3.py
import re

MUL = r'mul\((\d+),(\d+)\)'

with open("../_input/day3.txt", "r") as file:
    line = file.read()

segments = map(lambda s: s[0],
               map(lambda s: s.split('don\'t()'), line.split('do()')))

print(
    sum(
        map(lambda x: int(x[0]) * int(x[1]), re.findall(MUL, line))))
print(
    sum(
        map(lambda s: sum(
            map(lambda x: int(x[0]) * int(x[1]), re.findall(MUL, s))), segments)))

File: ruby/day3.rb
MUL = /mul\((\d+),(\d+)\)/.freeze

line = File.read("../_input/day3.txt")

segments = line.split("do()").map { |s| s.split("don't()") }.map(&:first)

puts line.scan(MUL).map { |a, b| a.to_i * b.to_i }.sum
puts segments.map { |s| s.scan(MUL).map { |a, b| a.to_i * b.to_i }.sum }.sum
